### PR TITLE
Properly update the threads table when thread events are redacted

### LINKED ISSUE #14241: /threads includes thread roots that have no thread attached to it anymore
## Description

/threads includes thread roots that have no thread attached to it anymore due to redactions.

## Steps to reproduce

1. create a thread
2. write replies
3. redact all replies

## Expected Behavior

The /threads endpoint stops including this thread in its response once all responses have been redacted

## Actual Behavior

The /threads endpoint still includes the now empty threads 

## Additional Information

The homeserver correctly removed the bundled relationship from the threads, yet it still includes the roots in the /threads response.

This issue is likely related to the [MSC3856 implementation](https://github.com/matrix-org/synapse/commit/3bbe532abb7bfc41467597731ac1a18c0331f539)

### PR DESCRIPTION
**Pull request by clokep**
When the last event in a thread was redacted we were not properly updating the `threads` table to either:

* Find the *new* latest event in the thread and store it into the table; or
* Removing the thread from the table if it is no longer a thread (i.e. all events in the thread were redacted).

Fixes #14241

### CODE REVIEW

File: synapse/storage/databases/main/events.py [hunk 1]
```diff
@@ -2081,7 +2081,9 @@ def _handle_redact_relations(
             txn.execute(sql, (redacted_relates_to, RelationTypes.THREAD))
 
             row = txn.fetchone()
-            # If a new latest event is found, update the threads table.
+            # If a latest event is found, update the threads table, this might
+            # be the same current latest event (if an earlier event in the thread
+            # was redacted).
```
Comment (DMRobertson) [inline synapse/storage/databases/main/events.py:2085]: I was going to ask on an earlier commit: could this be a no-op update? Answer: yes. But that's fine.

**Review by DMRobertson — APPROVED** [review]
(Still draft, sorry if this is premature.)

Seems reasonable to me (thank you for a neat commit history). I'd be okay to squeeze it into 1.70 but I'll let @erikjohnston have the final call there I think.

File: synapse/storage/databases/main/events.py [hunk 2]
```diff
@@ -2046,21 +2046,24 @@ def _handle_redact_relations(
...
+        if rel_type == RelationTypes.THREAD:
+            self.store._invalidate_cache_and_stream(
+                txn, self.store.get_thread_summary, (redacted_relates_to,)
+            )
+            self.store._invalidate_cache_and_stream(
+                txn, self.store.get_thread_participated, (redacted_relates_to,)
+            )
+            self.store._invalidate_cache_and_stream(
+                txn, self.store.get_threads, (room_id,)
+            )
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/databases/main/events.py:2070]: > Just want to sanity check here: is it really the case that these caches should be invalidated in separate groups based on the rel_type?

It should be:

1. `get_aggregation_groups_for_event` only returns data about `m.annotation` related events. (We should rename this at some point...)
2. `get_applicable_edit` only searches over `m.replace` events.
3. `get_thread_summary`, `get_thread_participated`, and `get_threads` are only updated if you create or redact an event with `m.thread` relation.

> E.g. suppose I haven't participated in a thread, then I add an annotation the root event. Am I still not considered as having participated in the thread?

Yes, "participated" is if you've sent the root event or a reply. Other relations to the root or threaded replies do not count.

File: synapse/storage/databases/main/events.py [hunk 3]
```diff
@@ -2065,10 +2069,6 @@ def _handle_redact_relations(
                 txn, self.store.get_threads, (room_id,)
             )
 
-        self.db_pool.simple_delete_txn(
-            txn, table="event_relations", keyvalues={"event_id": redacted_event_id}
-        )
-
```
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/databases/main/events.py:2063]: Err, no this isn't fixing anything of that sort; it just makes it so the query to look for a new latest event in the thread won't possibly find the event that is about to be deleted.

File: synapse/storage/databases/main/events.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/storage/databases/main/events.py:2070]: > Yes, "participated" is if you've sent the root event or a reply. Other relations to the root or threaded replies do not count.

Great, that's exactly the confirmation I was looking for. Thanks!

File: synapse/storage/databases/main/events.py (hunk 1 above)
Comment (clokep) [reply synapse/storage/databases/main/events.py:2085]: I couldn't figure out a way to make this "nicer". We could also pull the latest thread info from the `threads` table first and if it matches then no-op, but that doesn't seem to be any "better" than just upserting?

Comment (DMRobertson) [reply synapse/storage/databases/main/events.py:2085]: Yeah, unless you already have the relevant info in memory I think this is completely fine.

File: synapse/storage/databases/main/events.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/databases/main/events.py:2063]: No problem! I should have made the commit message clearer at least.

File: synapse/storage/databases/main/events.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/storage/databases/main/events.py:2070]: For reference, this is in [the spec](https://spec.matrix.org/latest/client-server-api/#server-side-aggregation-of-mthread-relationships) (in more technical terms):

> `current_user_participated` is `true` when the authenticated user is either:
>
> 1. The `sender` of the event receiving the bundle (they sent the thread root).
> 2. The `sender` of an event which references the thread root with a `rel_type` of `m.thread`.

**Comment by erikjohnston** [comment]
I'm also happy to put this in 1.70 if you're happy to do an RC2

**Pull request merged**